aprx.8.in
.TH aprx 8 "@DATEVERSION@"
.LO 8
.SH NAME
.B Aprx-2
\- An APRS iGate application with integrated Digipeater.
.SH SYNOPSIS
.B aprx
.RB [ \-d [ d [ d ]]]
.RB [ \-e ]
.RB [ \-i ]
.RB [ \-v ]
.RB [ \-V ]
.RB [ \-l " \fIsyslogfacilityname\fR]"
.RB [ \-f " \fI@CFGFILE@\fR]"
.SH DESCRIPTION
The
.B aprx
program is a special purpose Ham-radio application supplying
infrastructure services for APRS protocol use.
.PP
A more detailed manual is available at:
.br
http://ham.zmailer.org/oh2mqk/aprx/aprx-manual.pdf
.PP
.SH FEATURES
The Aprx begun as a receive-only APRS iGate application with minimum system
support technology requirements.
This version has also multi-port digipeater support, transmit iGate,
and experimental D-PRS-to-APRS RF/Rx-iGate.
.PP
.IP \(bu 3
The Aprx does not require machine to have any other software in it, than things
in UNIX standard libc. In particular no special AX.25 libraries at all, nor
widgets or even C++ runtime.
.IP \(bu 3
Important goal has been to keep R/W memory footprint as small as possible,
and on general purpose i386 Linux a single radio port iGate+digipeater is
now around 250 kB of R/W memory allocations.
.IP \(bu 3
Any UNIX (and UNIX like) platform should work for the Aprx, or be trivially ported.
.IP \(bu 3
The Aprx can listen "TNC2 monitor" and "KISS" speaking TNCs on any serial ports.
.IP \(bu 3
For Aprx the serial port can be ordinary host computer port, a USB serial port,
or remote port on a remote server behind the internet, like cisco router AUX
ports (port 4001, TCP STREAM without TELNET escapes.)
.IP \(bu 3
The Aprx does not require machine to have AX.25 protocol support internally!
(Thus this works also on e.g. Solaris and BSD machines without PF\_AX25 sockets.)
.IP \(bu 3
On Linux machine with kernel internal AX.25 protocol support,
the Aprx can listen on it with promiscuous mode and in order to use that,
the Aprx must be started as
.I root
user, and be configured to list interface callsigns that APRS packets are
coming in.
The AX.25 socket listening is not in itself configurable, it is always exists
in Linux systems, and related configuration parameters are ignored in other
platforms.
This socket listening does not need auxiliary "libax25" to function.
.IP \(bu 3
The Aprx program can be run without root privileges at least against remote
serial port servers.  One must change local serial port ownership or
access-groups (if any are used) to userid that runs the program and possibly
do several changes of file paths in configuration file beginning
with its location (startup parameter).
How that is done is up to the user or system integrator of this program.
.IP \(bu 3
The Aprx connects with one callsign-ssid pair to APRS-IS core for all received
radio ports.
.IP \(bu 3
The Aprx Rx-iGate knows that messages with following tokens in AX.25 VIA
fields are not to be relayed into APRS-IS network:
.RS 9
.B "RFONLY, NOGATE, TCPIP, TCPXX"
.RE
.IP \(bu 3
The Aprx Rx-iGate knows that following source address prefixes are bogus and
thus messages with them are to be junked:
.RS 9
.B "WIDE, RELAY, TRACE, TCPIP, TCPXX, NOCALL, N0CALL"
.RE
.IP \(bu 3
The Aprx Rx-iGate Drops all
.I query
messages ("?").
.IP \(bu 3
The Aprx Rx-iGate opens up all 3rd party messages ("}"), and checks the internal
data if it is OK to be gated out to APRS-IS.
.IP \(bu 3
The Aprx has built-in "Erlang monitor" mechanism that telemeters each receiving
interface to APRS-IS. It can also syslog the interface specific channel occupancy,
and optionally can output to STDOUT.
.IP \(bu 3
The Aprx (since version 1.91) can do digipeater functions.
.IP \(bu 3
The Aprx (since version 1.99) does have experimental D-STAR D-PRS to APRS
gateway functionality.  See the
.I aprx-manual.pdf
for details.
.IP \(bu 3
The Aprx can be run on systems without writable storage, even with very little
memory, like on NSLU2, and OpenWrt platforms.
The experiments have shown that a single radio Tx-iGate+digipeater works
with less than 300 kB of writable RAM for the Aprx itself.
Additional memory is necessary for operating system services of TCP/IP
networking, and serial port drivers.
.PP
.SH OPTIONS
The
.B aprx
has following runtime options:
.TP
.B "\-i"
Keep the program foreground without debugging outputs.
.TP
.B "\-d"
Turn on verbose debugging, outputs data to STDOUT.
.TP
.B "\-dd"
the "more debug" mode shows also details of network interaction with
the APRS-IS network service.
.TP
.B "\-ddd"
the "even more debug" mode shows also detail classification of
every kind of frame received in KISS variants.
.TP
.B "\-e"
.I "Erlang output"
prints 10 minute and 60 minute traffic accumulation byte counts, and guestimates
on channel occupancy, alias "Erlang".
These outputs are sent to STDOUT, which system operator may choose to log elsewere.
This is independent if the "\-l" option below.
.TP
.BI "\-f " "@CFGFILE@"
Configuration file, given path is built-in default, and can be overridden by the program runner.
.TP
.BR "\-l" " \fIsyslogfacilityname\fR"
Defines
.RB syslog (3)
facility code used by the erlang reporter by defining its name.
Default value is:
.BR NONE ,
and accepted values are:
.BR LOG_DAEMON ,
.BR LOG_FTP ,
.BR LOG_LPR ,
.BR LOG_MAIL ,
.BR LOG_NEWS ,
.BR LOG_USER ,
.BR LOG_UUCP ,
.BR LOG_LOCAL0 ,
.BR LOG_LOCAL1 ,
.BR LOG_LOCAL2 ,
.BR LOG_LOCAL3 ,
.BR LOG_LOCAL4 ,
.BR LOG_LOCAL5 ,
.BR LOG_LOCAL6 ,
.BR LOG_LOCAL7 .
That list is subject to actual facility code set in the system,
and in any case if you specify a code that is not known, then the program
will complain during the startup, and report it.
This is independent of the "\-e" option above.
.TP
.B "\-v"
Verbose logging of received traffic to STDOUT.
Lines begin with reception timestamp (UNIX time_t seconds), then TAB,
and either data as is, or with prefix byte: "*" for "discarded due to data content",
or possibly "#" for "discarded due to APRS-IS being unreachable".
.TP
.B "\-V"
Print source version compiled to this binary, and exit.
.PP
.SS DEBUGGING SYSTEM
Use parameter set 
.B "\-ddv"
(or
.BR "\-dddv" )
to test new configuration by running it synchronously to console.
.PP
.SS NORMAL OPERATION
Running the
.B aprx
program without any of option flags:
.BR "\-d" ,
.BR "\-v" ", or"
.B "\-e"
reads possibly given configuration, then automatically backgrounds the process, and writes
.IR pidfile .
When the process whose number written in
.I pidfile
is then sent a SIGTERM signal, it automatically shuts down itself, and removes the
.IR pidfile .
The
.I pidfile
can be runtime configured with the
.BI \-f " @CFGFILE@"
file, and it has default name of:
.IR "@VARRUN@/aprx.pid" .
.PP

.SH CONFIGURATION FILE
The configuration file is used to setup the program to do its job.
.PP
You can construct following configurations:
.PP
.IP \(bu 3
A
.I receive-only
iGate server.
.IP \(bu 3
A digipeater with bi-directional iGate server.
.IP \(bu 3
A
.I "single radio"
digipeater.  (The most common type of digipeater.)
.IP \(bu 3
A
.I multi-interfaced
digipeater relaying traffic in between multiple radios.  (On same or on separate frequencies.)
.IP \(bu 3
A
.I "viscuous digipeater,"
which relays a packet it heard from viscuous source after the viscuous delay,
.I unless it was heard more times than only once,
or it was heard from non-viscuous source before the viscuous one was digipeated.
This allows of making fill-in digipeaters that will not digipeat the packet,
if that same packet was heard twice or more before the viscuos delay expired.
.PP
In the configuration file a line ending backslash (\\) character concatenates
next input line into itself. Combined result can be up to 8000 bytes long.
This combination can be a bit surprising:
.RS 3em
.nf
\fC#beacon .... long text  \\
       continuation\fR
.fi
.RE
results in single long input line that begins with '#' (it is comment) and all
continuations following it have been folded in.
Presented line number of combined continuation is the line number of the
.I last
line segment in this type of multi-line input.
.PP
In the configuration file there is special treatment for quoted strings.
They are stripped of the outer quotes, and "\fC\\\fR" character is processed within
the source string to produce an output string.
The escapes are:
.TP
.B "\fC\\\\n"
Produces newline character (Control-J) on the output string.
.TP
.B "\fC\\\\r"
Produces carriage return character (Control-M) on the output string.
.TP
.B "\fC\\\\\\\\"
Places a back-slash on the output string.
.TP
.B "\fC\\\\""
.\" foo "
Places a double-quote on the output string.
.TP
.B "\fC\\\\'"
Places a single-quote on the output string.
.TP
.B "\fC\\\\xHH"
Lower-case "x" precedes two hex digits which ensemble is then converted to a single byte in the output string.
.PP
The complex encodings are for possible initstrings of the external devices,
and in particular for initstrings even a nul byte ( \\x00 ) is supported.
.PP
A configuration token without surrounding quotes does not understand the backslash escapes.
.PP
.nf
\fC
#
#  Sample configuration file for the APRX -- an Rx-only APRS iGate with
#  Digipeater functionality.
#
#
# Simple sample configuration file for the APRX-2
#
# This configuration is structured with Apache HTTPD style tags
# which then contain subsystem parameters.
#

#
# For simple case, you need to adjust 4 things:
#   - Mycall parameter
#   - Select correct type of interface (ax25-device or serial-device)
#   - Optionally set a beacon telling where this system is

ChangeLog
2014-10-21 Owen Duffy - VK2OMD

	*   Fixes to defects, mainly in digipeater code

	*   Security enhancement - runs as user

	*   Logging improvements

	*   New digi features: maxhops, lastrepeat

	*   v2.09 baseline

	*   Source at https://github.com/owenduffy/aprx

2014-08-29 Matti Aarnio - OH2MQK - KP20NG  <oh2mqk@sral.fi>

	* beacon.c:
	    Demote some nuisance logging to "-dd" level.

2014-03-31 Matti Aarnio - OH2MQK - KP20NG  <oh2mqk@sral.fi>

	* aprx.h, interface.c, igate.c, dprsgw.c:
	    Deliver all APRS-IS supplied tokens to the interface 3rd-party
	    receiver

	* beacon.c:
	    Limit read amount of 256 bytes, excess is rejected and usually
	    also the beaconing is skipped.

	* aprx.c, aprx.h, igate.c:
	    Moved rflog() from igate.c to aprx.c.

2014-03-24 Matti Aarnio - OH2MQK - KP20NG  <oh2mqk@sral.fi>

	* interface.c:
	    Igated 3rd-party frames are produced with different TNC2 format
	    message from AX.25 frame message.  This is because TNC2 format
	    is used at filter processing, and needs to have original source
	    address arriving from APRS-IS.

	* ttyreader.c:
	    Set default read timeout to 60 minutes, it can be configured to
	    any value from 1 second to 4 hours as channel busyness supports,
	    and even disabled.

	* ttyreader.c:
	    Use aprxlog() to record tty state changes: close/open

	* netresolver.c:
	    Use same "die_now" flag as all other thread loops do.

	* historydb.c:
	    Fix key pickup of sourcename.
	    Debug print key in history_db_insert_()

	* agwpesocket.c:
	    Code cleaning.

	* cellmalloc.c:
	    No need to include <pthread.h>

	* configure.in:
	    Make pthread autoconfig default with option to disable it.

	* pbuf.c:
	    Declare pbuf_alloc() static.

2014-03-22 Matti Aarnio - OH2MQK - KP20NG  <oh2mqk@sral.fi>

	* aprx.c:
	    Fix aprxlog() function internal varargs usage.
	    It must be reset for each vfprintf() call...

	* aprx.c:
	    Initial value of time_reset = 1; start in "reset state",
	    which does time resetting in all prepoll codes in main
	    loop. At second round and there after of the main loop
	    the reset_time flag will be reset after execution of all
	    prepoll functions.

	* aprsis.c:
	    Do initial connect at about current tick + 10 seconds.
	    Reset phase will also put next connect attempt at that time.

	* aprx.c:
	    Change stdout and stderr buffering definition to have
	    a buffer, and make line buffering in a way that works..

	* aprsis.c:
	    Correct calls to aprxlog().

	* configure.in, Makefile.in:
	    Revised the pthread support autoconfig to really find where
	    the libraries are, and set correct definitions and compile/link
	    time parameters.

2014-03-21 Matti Aarnio - OH2MQK - KP20NG  <oh2mqk@sral.fi>

	* aprx.h, interface.c, beacon.c, config.c, digipeater.c,
	  telemetry.c:
	    Change the "flags are bits in an int" to "packet bit fields"
	    allowing direct setting of flags.

	* aprx.c, beacon.c:
	    Do not use stdout FILE* to print things out of signal handler.
	    Use sprintf(3) to put things into local buffer, then use write(2)
	    to send them to file handle 1.  Otherwise GLIBC has some
	    interlock issue causing a deadlock in rare case. (Happened
	    with debug printouts in development.)

	* beacon.c:
	    Correct close() of exec type beacon file descriptor.

	* interface.c:
	    Correct primary <interface> flags defaults, and setting of
	    "telem-to-is <boolean>" and "telem-t-rf <boolean>" flags.

	* telemetry.c:
	    Debug-printout (debug>1) of telemetry about time until next
	    output, plus packet that is being telemetered along with
	    hex coding of interface flags. Also streamlined the telemetry
	    label output code.

2014-03-11 Matti Aarnio - OH2MQK - KP20NG  <oh2mqk@sral.fi>

	* ttyreader.c:
	    Do open /dev/ttySnnn up front in non-blocking mode so that
	    if the serial port hardware needs to be wired specially to
	    be considered active ( = flow control ) the serial port open
	    will not stop -- hang -- at it missing.

	* agwpesocket.c:
	    Update time comparison codes to current model.

	* aprx.c:
	    time_reset gets cleared only after first round through
	    the main loop.

	* beacon.c:
	    The 'file' and 'exec' directives need at most 256 bytes of read
	    buffer to get beacon body content. No need for 2kB buffer alloc.

2014-03-08 Matti Aarnio - OH2MQK - KP20NG  <oh2mqk@sral.fi>

	* digipeater.c:
	    Corrected the digipeating recognition algorithm.
	    If the current leading VIA field value without
	    H-bit set is not recognized as transmitter callsign,
	    transmitter alias, or value of <trace> or <wide>

aprx.c
/* **************************************************************** *
 *                                                                  *
 *  APRX -- 2nd generation APRS iGate and digi with                 *
 *          minimal requirement of esoteric facilities or           *
 *          libraries of any kind beyond UNIX system libc.          *
 *                                                                  *
 * (c) Matti Aarnio - OH2MQK,  2007-2014                            *
 *                                                                  *
 * **************************************************************** */
#include "aprx.h"

/* Bits used only in the main program.. */
#include <signal.h>
#ifdef HAVE_SYS_TIME_H
# include <sys/time.h>
#endif
#ifdef HAVE_TIME_H
# include <time.h>
#endif
#include <fcntl.h>
#ifdef HAVE_SYS_WAIT_H
#include <sys/wait.h>
#endif

int debug;
int verbout;
int erlangout;
const char *rflogfile;
const char *aprxlogfile;
const char *mycall;
float myloc_lat;
float myloc_coslat;
float myloc_lon;
const char *myloc_latstr;
const char *myloc_lonstr;

const char *tocall = "APRX29";

#ifndef CFGFILE
#define CFGFILE "/etc/aprx.conf"
#endif

const char *pidfile = VARRUN "/aprx.pid";

int die_now;
int log_aprsis;

const char *swname = "aprx";
const char *swversion = APRXVERSION;


static void sig_handler(int sig)
{
	die_now = 1;
	signal(sig, sig_handler);
        aprxlog("aprx ending (SIG %d) - %s",sig,swversion);
	if (debug) {
          // Avoid stdio FILE* interlocks within signal handler
          char buf[64];
	  sprintf(buf, "SIGNAL %d - DYING!\n", sig);
          write(1, buf, strlen(buf));
        }
}

static void sig_child(int sig)
{
	int status;
        int pid;
        while ((pid = waitpid(-1, &status, WNOHANG)) > 0) {
        	beacon_childexit(pid);
        }
}  

static void usage(void)
{
	printf("aprx: [-d[d[d]]][-e][-i][-v][-L][-l logfacility] [-f %s]\n",
	       CFGFILE);
	printf("    version: %s\n", swversion);
	printf("    -f %s:  where the configuration is\n", CFGFILE);
	printf("    -v:  Outputs textual format of received packets, and data on STDOUT.\n");
	printf("    -e:  Outputs raw ERLANG-report lines on SYSLOG.\n");
	printf("    -i:  Keep the program foreground without debugging printouts.\n");
	printf("    -l ...: sets syslog FACILITY code for erlang reports, default: LOG_DAEMON\n");
	printf("    -d:  turn debug printout on, use to verify config file!\n");
	printf("         twice: prints also interaction with aprs-is system..\n");
	printf("    -L:  Log also all of APRS-IS traffic on relevant log.\n");
	exit(64);		/* EX_USAGE */
}

static void versionprint()
{
	printf("aprx: %s\n", swversion);
        exit(1);
}

void fd_nonblockingmode(int fd)
{
	int __i = fcntl(fd, F_GETFL, 0);
	if (__i >= 0) {
		/* set up non-blocking I/O */
		__i |= O_NONBLOCK;
		__i = fcntl(fd, F_SETFL, __i);
	}
	// return __i;
}

int time_reset = 1;             // observed time jump, initially as "reset is happening!"
static struct timeval old_tick; // monotonic
// static struct timeval old_now;  // wall-clock

static int timetick_count;

void timetick(void)
{
	++timetick_count;
        old_tick  = tick;
        //old_now = now;

	// Monotonic (or as near as possible) clock..
	// .. which is NOT wall clock time.
#ifdef HAVE_CLOCK_GETTIME
	struct timespec ts;
	clock_gettime(CLOCK_MONOTONIC, &ts);
	tick.tv_usec = ts.tv_nsec/1000;
	tick.tv_sec  = ts.tv_sec;
        // if (debug) printf("newtick: %d.%6d\n", tick.tv_sec, tick.tv_usec);
#else
	gettimeofday(&tick, NULL); // fallback when no clock_gettime() is available
#endif
        // Wall clock time
        // gettimeofday(&tick, NULL);

        // Main program clears this when appropriate
        int delta = 0;
        if (old_tick.tv_sec != 0) {
          delta = tv_timerdelta_millis(&old_tick, &tick);
          if (delta < -1) { // Up to 0.99999 seconds backwards for a leap second
            if (debug) {
              printf("MONOTONIC TIME JUMPED BACK BY %g SECONDS. ttcallcount=%d\n", delta/1000.0, timetick_count);
            }
            time_reset = 1;
          } else if (delta > 32000) { // 30.0 + leap second + margin
            if (debug) {
              printf("MONOTONIC TIME JUMPED FORWARD BY %g SECONDS. ttcallcount=%d mypid=%d\n", delta/1000.0, timetick_count, getpid());
            }
            time_reset = 1;
          } else {
            // Time is OK.
            // time_reset = 0;
          }
        } else {
          time_reset = 1;
          // This happens before argv is parsed, thus debug is never set.
          // But if it sets happens afterwards...
          if (debug) printf("Initializing MONOTONIC time\n");
        }
        // if (debug>1) printf("TIMETICK %ld:%6d  %d delta=%d ms\n", tick.tv_sec, tick.tv_usec, timetick_count, delta);
}

int main(int argc, char *const argv[])
{
	int i;
	const char *cfgfile = "/etc/aprx.conf";
	const char *syslog_facility = "NONE";
	int foreground = 0;
        int millis;
        int can_clear_timereset;

	/* Init the poll(2) descriptor array */
	struct aprxpolls app = APRXPOLLS_INIT;

        timetick(); // init global time references

        setvbuf(stdout, NULL, _IOLBF, BUFSIZ);
        setvbuf(stderr, NULL, _IOLBF, BUFSIZ);

	while ((i = getopt(argc, argv, "def:hiLl:vV?")) != -1) {
		switch (i) {
		case '?':
		case 'h':
			usage();
			break;
		case 'd':
			++debug;
			++foreground;
			break;
		case 'e':
			++erlangout;
			++foreground;
			break;
		case 'i':
			++foreground;
			break;
		case 'L':
			log_aprsis = 1;
			break;
		case 'l':
			syslog_facility = optarg;
			break;
		case 'v':
			++verbout;
			++foreground;
			break;
		case 'f':
			cfgfile = optarg;
			break;
                case 'V':
                	versionprint();
                	break;
		default:
			break;
		}
	}

	interface_init(); // before any interface system and aprsis init !
	erlang_init(syslog_facility);
	ttyreader_init();
#ifdef PF_AX25			/* PF_AX25 exists -- highly likely a Linux system ! */
	netax25_init();
#endif
#ifdef ENABLE_AGWPE
	agwpe_init();
#endif
	dupecheck_init(); // before aprsis_init() !
#ifndef DISABLE_IGATE
	aprsis_init();
#endif
	filter_init();
	pbuf_init();

	i = readconfig(cfgfile);
	if (i) {
	  fflush(stdout);
	  fprintf(stderr, "Seen configuration errors. Aborting!\n");
	  fflush(stderr);
	  exit(1); // CONFIION ERRORS SEEN! ABORT!
	}

	erlang_start(1);
#ifndef DISABLE_IGATE
	historydb_init();
#endif

	if (debug || verbout) {
	  if (!mycall
#ifndef DISABLE_IGATE
	      && !aprsis_login
#endif
	      ) {
		fflush(stdout);
		fprintf(stderr,
			"APRX: NO GLOBAL  MYCALL=  PARAMETER CONFIGURED, WILL NOT CONNECT APRS-IS\n(This is OK, if no connection to APRS-IS is needed.)\n");
	  } else if (!mycall
#ifndef DISABLE_IGATE
		     && !aprsis_login
#endif
		     ) {
		fflush(stdout);
		fprintf(stderr,
			"APRX: NO GLOBAL  APRSIS-LOGIN=  PARAMETER CONFIGURED, WILL NOT CONNECT APRS-IS\n(This is OK, if no connection to APRS-IS is needed.)\n");
	  }
	}

	if (!foreground) {
		/* See if pidfile exists ? */
		FILE *pf = fopen(pidfile, "r");
		if (pf) {	/* See if the pid exists ? */
			int rc, er;
			int pid = -1;
			fscanf(pf, "%d", &pid);
			fclose(pf);

			if (pid > 0) {
				rc = kill(pid, 0);
				er = errno;

				if ((rc == 0) || (er == EPERM)) {
					fflush(stdout);
					fprintf(stderr,
						"APRX: PIDFILE '%s' EXISTS, AND PROCESSID %d INDICATED THERE EXISTS TOO. FURTHER INSTANCES CAN ONLY BE RUN ON FOREGROUND!\n",
						pidfile, pid);
					fflush(stderr);
					exit(1);
				}
			}
		}
	}


	if (!foreground) {
		int pid = fork();
		if (pid > 0) {
			/* This is parent */
			exit(0);
		}
		/* child and error cases continue on main program.. */
		poll((void*)&pid, 0, 500);

	}

	if (1) {
		/* Open the pidfile, if you can.. */

		FILE *pf = fopen(pidfile, "w");

		setsid();	/* Happens or not ... */

		if (!pf) {
			/* Could not open pidfile! */
			fflush(stdout);
			fprintf(stderr, "COULD NOT OPEN PIDFILE: '%s'\n",
				pidfile);
			pidfile = NULL;
		} else {
			int f = fileno(pf);
			if (flock(f, LOCK_EX|LOCK_NB) < 0) {
				if (errno == EWOULDBLOCK) {
					printf("Could not lock pid file file %s, another process has a lock on it. Another process running - bailing out.\n", pidfile);
				} else {
					printf("Failed to lock pid file %s: %s\n", pidfile, strerror(errno));
				}
				exit(1);
			}
			
			fprintf(pf, "%ld\n", (long) getpid());
			// Leave it open - flock will prevent double-activation
			dup(f); // don't care what the fd number is
			fclose(pf);
		}
	}


	erlang_start(2);	// reset PID, etc..

	// Do following as late as possible..

	// In all cases we close STDIN/FD=0..
	// .. and replace it with reading from /dev/null..
	i = open("/dev/null", O_RDONLY, 0);
	if (i >= 0) { dup2(i, 0); close(i); }
	
	// Leave STDOUT and STDERR open

	if (!foreground) {
	  // when daemoning, we close also stdout and stderr..
	  dup2(0, 1);
	  dup2(0, 2);
	}

	// .. but not latter than this.


	// Set default signal handling

	signal(SIGTERM, sig_handler);
	signal(SIGINT,  sig_handler);
	signal(SIGHUP,  sig_handler);
	signal(SIGPIPE, SIG_IGN);
	signal(SIGCHLD, sig_child);

	// Must be after config reading ...
	netresolv_start();
#ifndef DISABLE_IGATE
	aprsis_start();
#endif
#ifdef PF_AX25			/* PF_AX25 exists -- highly likely a Linux system ! */
	netax25_start();
#endif
#ifdef ENABLE_AGWPE
	agwpe_start();
#endif
	telemetry_start();
#ifndef DISABLE_IGATE
	igate_start();
#endif

        aprxlog("aprx start - %s",swversion);

	// The main loop

        can_clear_timereset = 0;

	while (!die_now) {

        	timetick(); // pre-poll

		aprxpolls_reset(&app);
                tv_timeradd_millis( &app.next_timeout, &tick, 30000 ); // 30 seconds

		i = ttyreader_prepoll(&app);
                // if (debug>3)printf("after ttyreader prepoll - timeout millis=%d\n",aprxpolls_millis(&app));
#ifndef DISABLE_IGATE
		i = aprsis_prepoll(&app);
                // if (debug>3)printf("after aprsis prepoll - timeout millis=%d\n",aprxpolls_millis(&app));
#endif
		i = beacon_prepoll(&app);
                // if (debug>3)printf("after beacon prepoll - timeout millis=%d\n",aprxpolls_millis(&app));
#ifdef PF_AX25			/* PF_AX25 exists -- highly likely a Linux system ! */
		i = netax25_prepoll(&app);
                // if (debug>3)printf("after netax25 prepoll - timeout millis=%d\n",aprxpolls_millis(&app));
#endif
#ifdef ENABLE_AGWPE
		i = agwpe_prepoll(&app);
                // if (debug>3)printf("after agwpe prepoll - timeout millis=%d\n",aprxpolls_millis(&app));
#endif
		i = erlang_prepoll(&app);
                // if (debug>3)printf("after erlang prepoll - timeout millis=%d\n",aprxpolls_millis(&app));
		i = telemetry_prepoll(&app);
                // if (debug>3)printf("after telemetry prepoll - timeout millis=%d\n",aprxpolls_millis(&app));
		i = dupecheck_prepoll(&app);
                // if (debug>3)printf("after dupecheck prepoll - timeout millis=%d\n",aprxpolls_millis(&app));
		i = digipeater_prepoll(&app);
                // if (debug>3)printf("after digipeater prepoll - timeout millis=%d\n",aprxpolls_millis(&app));
#ifndef DISABLE_IGATE
		i = historydb_prepoll(&app);
                // if (debug>3)printf("after historydb prepoll - timeout millis=%d\n",aprxpolls_millis(&app));
		i = dprsgw_prepoll(&app);
                // if (debug>3)printf("after dprsgw prepoll - timeout millis=%d\n",aprxpolls_millis(&app));
#endif

                // All pre-polls are done
                if (can_clear_timereset) {
                  // if (time_reset) {
                  //   printf("Clearing time_reset.\n");
                  // }
                  time_reset = 0;
                } else {
                  can_clear_timereset = 1;
                }

		// if (app.next_timeout <= now.tv_sec)
                // app.next_timeout = now.tv_sec + 1;	// Just to be on safe side..

                millis = aprxpolls_millis(&app);
                if (millis < 10)
                  millis = 10;

		i = poll(app.polls, app.pollcount, millis);
                timetick(); // post-poll


		i = beacon_postpoll(&app);
		i = ttyreader_postpoll(&app);
#ifdef PF_AX25			/* PF_AX25 exists -- highly likely a Linux system ! */
		i = netax25_postpoll(&app);
#endif
#ifdef ENABLE_AGWPE
		i = agwpe_postpoll(&app);
#endif
#ifndef DISABLE_IGATE
		i = aprsis_postpoll(&app);
#endif
		i = erlang_postpoll(&app);
		i = telemetry_postpoll(&app);
		i = dupecheck_postpoll(&app);
		i = digipeater_postpoll(&app);
#ifndef DISABLE_IGATE
		i = historydb_postpoll(&app);
		i = dprsgw_postpoll(&app);
#endif

	}
	aprxpolls_free(&app); // valgrind..

#ifndef DISABLE_IGATE
	aprsis_stop();
#endif
	netresolv_stop();

	if (pidfile) {
		unlink(pidfile);
	}

	exit(0);
}


void printtime(char *buf, int buflen)
{
	struct timeval tv;
	struct tm t;

        // Wall lock time for printouts
	gettimeofday(&tv, NULL);
        gmtime_r(&tv.tv_sec, &t);
	// strftime(timebuf, 60, "%Y-%m-%d %H:%M:%S", t);
	sprintf(buf, "%04d-%02d-%02d %02d:%02d:%02d.%03d",
		t.tm_year+1900,t.tm_mon+1,t.tm_mday,
		t.tm_hour,t.tm_min,t.tm_sec,
		(int)(tv.tv_usec / 1000));
}

static struct syslog_facs {
	const char *name;
	int fac_code;
} syslog_facs[] = {
	{
	"NONE", -1}, {
	"LOG_DAEMON", LOG_DAEMON},
#ifdef LOG_FTP
	{
	"LOG_FTP", LOG_FTP},
#endif
#ifdef LOG_LPR
	{
	"LOG_LPR", LOG_LPR},
#endif
#ifdef LOG_MAIL
	{
	"LOG_MAIL", LOG_MAIL},
#endif
#ifdef LOG_USER
	{
	"LOG_USER", LOG_USER},
#endif
#ifdef LOG_UUCP
	{
	"LOG_UUCP", LOG_UUCP},
#endif
	{
	"LOG_LOCAL0", LOG_LOCAL0}, {
	"LOG_LOCAL1", LOG_LOCAL1}, {
	"LOG_LOCAL2", LOG_LOCAL2}, {
	"LOG_LOCAL3", LOG_LOCAL3}, {
	"LOG_LOCAL4", LOG_LOCAL4}, {
	"LOG_LOCAL5", LOG_LOCAL5}, {
	"LOG_LOCAL6", LOG_LOCAL6}, {
	"LOG_LOCAL7", LOG_LOCAL7}, {
	NULL, 0}
};

void aprx_syslog_init(const char *syslog_facility_name)
{
	static int done_once = 0;
	int syslog_fac = LOG_DAEMON, i;

	if (done_once) {
		closelog();	/* We reconfigure from config file! */
	} else
		++done_once;
	for (i = 0;; ++i) {
		if (syslog_facs[i].name == NULL) {
			fprintf(stderr,
				"Sorry, unknown erlang syslog facility code name: %s, not supported in this system.\n",
				syslog_facility_name);
			fprintf(stderr, "Accepted list is:");
			for (i = 0;; ++i) {
				if (syslog_facs[i].name == NULL)
					break;
				fprintf(stderr, " %s",
					syslog_facs[i].name);
			}
			fprintf(stderr, "\n");
			break;
		}
		if (strcasecmp(syslog_facs[i].name, syslog_facility_name)
		    == 0) {
			syslog_fac = syslog_facs[i].fac_code;
			break;
		}
	}

	if (syslog_fac >= 0) {
		erlangsyslog = 1;
		openlog("aprx", LOG_NDELAY | LOG_PID, syslog_fac);
	}
}

#ifdef HAVE_STDARG_H
#ifdef __STDC__
void aprxlog(const char *fmt, ...)
#else
void aprxlog(fmt)
#endif
#else
/* VARARGS */
void aprxlog(va_list)
va_dcl
#endif
{
	va_list ap;
	char timebuf[60];

        printtime(timebuf, sizeof(timebuf));
	if (verbout) {
#ifdef 	HAVE_STDARG_H
          va_start(ap, fmt);
#else
          const char *fmt;
          va_start(ap);
          fmt    = va_arg(ap, const char *);
#endif

	  fprintf(stdout, "%s ", timebuf);
	  vfprintf(stdout, fmt, ap);
          (void)fprintf(stdout, "\n");

#ifdef 	HAVE_STDARG_H
          va_end(ap);
#endif
	}

        if (aprxlogfile) {
          FILE *fp;

#ifdef 	HAVE_STDARG_H
          va_start(ap, fmt);
#else
          const char *fmt;
          va_start(ap);
          fmt    = va_arg(ap, const char *);
#endif


#if defined(HAVE_PTHREAD_CREATE) && defined(ENABLE_PTHREAD)
          pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, NULL);
#endif
          fp = fopen(aprxlogfile, "a");
          if (fp != NULL) {
            setlinebuf(fp);
            fprintf(fp, "%s ", timebuf);
            vfprintf(fp, fmt, ap);
            (void)fprintf(fp, "\n");
            fclose(fp);
          }
#if defined(HAVE_PTHREAD_CREATE) && defined(ENABLE_PTHREAD)
          pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
#endif

#ifdef 	HAVE_STDARG_H
          va_end(ap);
#endif
        }
}


/* ---------------------------------------------------------- */

void rfloghex(const char *portname, char direction, int discard, const uint8_t *buf, int buflen)
{
}

void rflog(const char *portname, char direction, int discard, const char *tnc2buf, int tnc2len)
{
	if (rflogfile) {
#if defined(HAVE_PTHREAD_CREATE) && defined(ENABLE_PTHREAD)
        	pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, NULL);
#endif

		FILE *fp = NULL;
                const char *p;
		if (strcmp("-",rflogfile)==0) {
		  if (debug < 2) return;
		  fp = stdout;
		} else {
		  fp = fopen(rflogfile, "a");
		}
		
		if (fp) {
		  char timebuf[60];
		  printtime(timebuf, sizeof(timebuf));
	  
		  (void)fprintf(fp, "%s %-9s ", timebuf, portname);
		  (void)fprintf(fp, "%c ", direction);

		  if (discard < 0) {
		    fprintf(fp, "*");
		  }
		  if (discard > 0) {
		    fprintf(fp, "#");
		  }
                  //replace non printing TNC2 characters in log print
                  for(p=tnc2buf;p<tnc2buf+tnc2len;p++){
                    if(*p<0x20 || *p>0x7e)
                      fprintf(fp,"<0x%02x>",*p);
                    else
                      fputc(*p,fp);
                  }
                  fputc('\n',fp);

		  if (fp != stdout)
		    fclose(fp);
		}
#if defined(HAVE_PTHREAD_CREATE) && defined(ENABLE_PTHREAD)
		pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
#endif
	}
}
